tooling/helpers/range.go

```diff
@@ -0,0 +1,176 @@
+func (b ByteRange) getRangeStringWithoutPrefix(t *testing.T) string {
+	if b.To == nil {
+		return fmt.Sprintf("%d-", b.From)
+	}
+
+	to := *b.To
+	if to >= 0 {
+		return fmt.Sprintf("%d-%d", b.From, to)
+	}
+
+	if to < 0 && b.From != 0 {
+		t.Fatalf("for a suffix request the From field must be 0")
+	}
+	return fmt.Sprintf("%d", to)
+}
+
+func (b ByteRange) getRange(t *testing.T, totalSize int64) (uint64, uint64) {
+	if totalSize < 0 {
+		t.Fatalf("total size must be greater than 0")
+	}
+
+	if b.To == nil {
+		return b.From, uint64(totalSize)
+	}
+
+	to := *b.To
+	if to >= 0 {
+		return b.From, uint64(to)
+	}
+
+	if to < 0 && b.From != 0 {
+		t.Fatalf("for a suffix request the From field must be 0")
+	}
+
+	start := int64(totalSize) + to
+	if start < 0 {
+		t.Fatalf("suffix request must not start before the start of the file")
+	}
+
+	return uint64(start), uint64(totalSize)
+}
+
+type ByteRanges []ByteRange
+
+func (b ByteRanges) GetRangeString(t *testing.T) string {
+	var rangeStrs []string
+	for _, r := range b {
+		rangeStrs = append(rangeStrs, r.getRangeStringWithoutPrefix(t))
+	}
+	return fmt.Sprintf("bytes=%s", strings.Join(rangeStrs, ","))
+}
+
+// SingleRangeTestTransform takes a test where there is no "Range" header set in the request, or checks on the
+// StatusCode, Body, or Content-Range headers and verifies whether a valid response is given for the requested range.
+//
+// Note: HTTP Range requests can be validly responded with either the full data, or the requested partial data
+func SingleRangeTestTransform(t *testing.T, baseTest test.SugarTest, brange ByteRange, fullData []byte) test.SugarTest {
+	modifiedRequest := baseTest.Request.Clone().Header("Range", brange.GetRangeString(t))
+	if baseTest.Requests != nil {
+		t.Fatal("does not support multiple requests or responses")
+	}
+
+	fullSize := int64(len(fullData))
+	start, end := brange.getRange(t, fullSize)
+
+	rangeTest := test.SugarTest{
+		Name:     baseTest.Name,
+		Hint:     baseTest.Hint,
+		Request:  modifiedRequest,
+		Requests: nil,
+		Response: test.AllOf(
+			baseTest.Response,
+			test.AnyOf(
+				test.Expect().Status(http.StatusPartialContent).Body(brange.RangeBytes).Headers(
+					test.Header("Content-Range").Equals("bytes {{start}}-{{end}}/{{length}}", start, end, fullSize),
+				),
+				test.Expect().Status(http.StatusOK).Body(fullData),
+			),
+		),
+		Responses: baseTest.Responses,
+	}
+
+	return rangeTest
+}
+
+// MultiRangeTestTransform takes a test where there is no "Range" header set in the request, or checks on the
+// StatusCode, Body, or Content-Range headers and verifies whether a valid response is given for the requested ranges.
+//
+// Note: HTTP Multi Range requests can be validly responded with one of the full data, the partial data from the first
+// range, or the partial data from all the requested ranges
+func MultiRangeTestTransform(t *testing.T, testWithoutRangeRequestHeader test.SugarTest, branges ByteRanges, fullData []byte) test.SugarTest {
```

Comment:
Note: We could have a combined function for Single + Multi ranges but seems like it'd require implementing `Clone` for Responses.

We can do `Clone` for responses if we want (I've got some code locally from an attempt), but the clones are not completely deep because we pass functions around for checks. Thoughts?

Reply:
You thought about this more than I did, why do we need Clone?
I'd assume the Single case is "just" the Many case with N = 1.

Worst case scenario, I guess we could even do something like this:

```go
func WithRangesCheck(t *testing.T, test test.SugarTest, branges ByteRanges, fullData []byte) {
    if (branges.length == 0) {
      panic()
    }
    else if (branges.length == 1) {
       return singleRangeTransform(...)
    }
    else {
       return multipleRangeTransform(...)
    }
}

// return a ByteRangeS to please type checker AND users
func SimpleByteRange(from, to uint64, data []byte) ByteRanges {
   ...
}

func ManyByteRanges(...ByteRanges) {
   ...
}
```

Thinking about usage & consistency sugar API we could even do:

```go
ByteRange(1, 2, bytes)
  .And(5, 6, bytes)
  .And(8, 12, bytes)
```

Regarding Clone:
the `CheckFunc` is more of a temporary bandaid for cases that are not yet supported by the Sugar API. I'd avoid designing around it. I'd implement the `Clone()` function as a deep clone and assume we'll never need shallow cloning. It's easier to relax constraints later than add them. We could provide a `CheckPureFunc` for internal checks, and _trustworthy_ users

Reply:
> Worst case scenario, I guess we could even do something like this:

What I was thinking was that we could make a wrapper where the user passes the test for a standard non-range request and we could return 2 tests `(singleRange, multiRange)` (or maybe 3 if we include `noRange`) and so make it easier to test for ranges basically anywhere we wanted to.